### ISSUE TITLE
user deleting himself - should it delete all his account_memberships? accounts?

### ISSUE THREAD

**Issue by yshmarov** [yshmarov/moneygun#106]
Currently user can delete his account. It is tested here: https://github.com/yshmarov/askvote/blob/ffab700daef8a7ffc9cc06fba499b2ea0fc186eb/test/models/user_test.rb#L0-L1

I think:
- User with account_memberships that have downstream associated objects can only be soft deleted
- If a user is soft deleted, membership should reflect that user has no system access 
- If user is the only admin of an Account, Account can be deleted, User can be deleted
- If Account has an active subscription, it can not be deleted

Generally leaving an Account makes sense.
Deleting a User does not.

****

**a user deleting himself should not delete memberships. member should have some sort of state "without user"**

**Comment by mwala-zm** [comment]
I was thinking, a soft delete may be good in the event the user want to reactivate their account. Maybe they want to have the same account email for billing reasons.
Applying a soft delete (Archive) to users that are members (Admin or member) of an organisation or some sort of deactivation. This will keep memberships for easier reactivation.
But, we can still allow users the option to delete their account ensuring that they have no memberships or any other associations.

**Comment by yshmarov** [comment]
> I was thinking, a soft delete may be good in the event the user want to reactivate their account. Maybe they want to have the same account email for billing reasons. Applying a soft delete (Archive) to users that are members (Admin or member) of an organisation or some sort of deactivation. This will keep memberships for easier reactivation. But, we can still allow users the option to delete their account ensuring that they have no memberships or any other associations.

yep, Slack for example does not delete memberships. It just deactivates them.

I think we can safely delete membership if it has no associations. If it does have associations - archive (unassociate user).

```
# models/membership.rb
  has_many :projects, dependent: :restrict_with_error

  scope :archived, -> { where(user_id: nil) }

  def archived?
    user_id.nil?
  end
```

I think users should be able to completely delete their account (right to be forgotten, gdpr, etc). I don't want to implement archiving User in this app. It feels like an edge case/feature that would be hardly ever used.

**Comment by yshmarov** [comment]
will abstract lifecycle from my prod app later
